charger: add support for TI BQ25620

parthitce opened the pull request:
Add support for Texas Instruments BQ25620 charger, which has the capability to handle upto 3.5A. As the device supports additional features which aren't part of current charger API, following API's introduced.

- get and set multiple properties with `charger_get_props` and `charger_set_props` with array of `union charger_propval`
- USB type identification based on battery charger specification. Heavily inherited from Linux Kernel values
- Additional properties like charger & batter current and voltage. Although these belongs to fuel gauge, charger of this kind supports reading these values with built-in ADC.
- Adds additional health property for dead battery
- Information properties model name and manufacturer information

Signed-off-by: Parthiban Nallathambi <parthiban@linumiz.com>

parthitce committed:
charger: add battery health dead property

based on battery current, battery voltage values internal ADC's
in chargers like BQ25620/BQ25622 tracks dead batteries. Add
support to new health status as dead.

Signed-off-by: Parthiban Nallathambi <parthiban@linumiz.com>

parthitce committed:
charger: add method to read and write multiple properties

As like fuel guage, charger properties reading and writing can also
be possible to do multiple. Add a new interface API/syscall to read/
write multiple properties using array.

Signed-off-by: Parthiban Nallathambi <parthiban@linumiz.com>

parthitce committed:
charger: add battery present voltage and current property

chargers like BQ25620/BQ25622 supports a way to get the battery's
present voltage and current values. Add property to support it.

Signed-off-by: Parthiban Nallathambi <parthiban@linumiz.com>

rriveramcrus commented:
Added the USB label here to indicate that there may be some discussion here related to how to implement the BC1.2 functionality of this chip.

rriveramcrus commented on drivers/charger/charger_bq2562x.c:
```diff
@@ -0,0 +1,1035 @@
+
+static int bq2562x_hw_init(const struct device *dev)
+{
+	int ret;
+	struct bq2562x_data *data = dev->data;
+	const struct bq2562x_config *const config = dev->config;
```
Reverse pyramid declarations when possible please.

parthitce replied:
meaning longer form to shorter form i.e config , data and ret in this case?

rriveramcrus replied:
correct :)

https://hisham.hm/2018/06/16/when-listing-repeated-things-make-pyramids/
Some additional context on *why* if you are curious.

rriveramcrus commented on drivers/charger/charger_bq2562x.c:
```diff
@@ -0,0 +1,1035 @@
+		*charge_status = CHARGER_STATUS_DISCHARGING;
+	} else if (!status) {
+		*charge_status = CHARGER_STATUS_NOT_CHARGING;
+
+		ret = bq2562x_get_chrg_volt(dev, &chrg_volt);
+		if (ret < 0) {
+			return 0;
+		}
+
+		ret = bq2562x_get_vbat_adc(dev, &vbat_adc);
+		if (ret) {
+			return 0;
+		}
+
+		if (chrg_volt >= (vbat_adc - BQ2562X_ADC_VBAT_STEP_uV)) {
+			*charge_status = CHARGER_STATUS_FULL;
+		}
```
I don't love this idea of emulating a charge termination indicator bit in the driver. Is this a must-have for your system or can we just remove this?

parthitce replied:
Good question, technically this needs to come from Fuel Gauge. I can remove that part as it's not stable when tested with different sizes of battery. Thanks.

rriveramcrus commented on dts/bindings/charger/ti,bq2562x.yaml:
```diff
@@ -0,0 +1,108 @@
+      - 1
+    description: |
+      Charger control thermal regulation threshold
+      0: 60 degree celcius
+      1: 120 degree celcius
```
typo here, celcius should be celsius

rriveramcrus commented on dts/bindings/charger/ti,bq2562x.yaml:
```diff
@@ -0,0 +1,108 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  BQ25620/BQ25622 is a highly integrated switch-mode buck-boost charger for 1-4 cell
+  Li-ion batteries and Li-polymer batteries. The device charges a battery from a
+  wide range of input sources including legacy USB adapters to high voltage USB
+  PD adapters and traditional barrel adapters.
```
I think we can reduce this down to just one sentence. Also this charger is not multi-cell capable, only single cell.

rriveramcrus commented on drivers/charger/charger_bq2562x.c:
```diff
@@ -0,0 +1,1035 @@
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/util.h>
+
+LOG_MODULE_REGISTER(ti_bq24190);
```
Use the `CONFIG_CHARGER_LOG_LEVEL` like so `LOG_MODULE_REGISTER(ti_bq24190, CONFIG_CHARGER_LOG_LEVEL);`

rriveramcrus commented on drivers/charger/charger_bq2562x.c:
```diff
@@ -0,0 +1,1035 @@
+	}
+
+	charger_enable = chrg_ctrl_0 & BQ2562X_CHRG_EN;
+	if (charger_enable) {
```
https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html

This breaks MISRA-C 2012 [Rule 14.4](https://gitlab.com/MISRA/MISRA-C/MISRA-C-2012/Example-Suite/-/blob/master/R_14_04.c#L44)

rriveramcrus replied:
You still have a lot of these everywhere. I'm not going to point each one out, but the idea here is that every conditional should explicitly resolve to a boolean. 

```c
/* not okay */
if (ret)

/* fair game */
if (ret < 0)

/* fair game */
if (ret != 0)
```

rriveramcrus commented on drivers/charger/charger_bq2562x.c:
```diff
@@ -0,0 +1,1035 @@
+	chrg_curr = CLAMP(chrg_curr, BQ2562X_ICHG_I_MIN_uA,
+					chrg_curr_max);
+
+	bq2562x_set_charge_enable(dev, 0);
```
This breaks [Dir 4.7](https://gitlab.com/MISRA/MISRA-C/MISRA-C-2012/Example-Suite/-/blob/master/D_04_07.c)

rriveramcrus commented on drivers/charger/charger_bq2562x.h:
```diff
@@ -0,0 +1,221 @@
+#define BQ2562X_PART_INFO		0x38
+
+/* REG0x02_Charge_Current_Limit */
+#define BQ2562X_ICHG_I_MIN_uA		80000
```
Lowercase letters breaks the syntax highlighting on here. :(

rriveramcrus commented on drivers/charger/charger_bq2562x.c:
```diff
@@ -0,0 +1,1035 @@
+{
+	const struct bq2562x_config *const config = dev->config;
+	struct bq2562x_data *data = dev->data;
+	int chrg_volt_max = data->constant_charge_voltage_max_uv;
```
`chrg_volt_max` should be a `uint32_t`

rriveramcrus commented on drivers/charger/charger_bq2562x.c:
```diff
@@ -0,0 +1,1035 @@
+	return 0;
+}
+
+static int bq2562x_set_chrg_volt(const struct device *dev, int chrg_volt)
```
`chrg_volt` should be `uint32_t`

rriveramcrus commented on drivers/charger/charger_bq2562x.h:
```diff
@@ -0,0 +1,221 @@
+#define BQ2562X_ICHG_I_MIN_uA		80000
+#define BQ2562X_ICHG_I_MAX_uA		4000000
+#define BQ2562X_ICHG_I_DEF_uA		1040000
+#define BQ2562X_ICHG_I_MOVE_STEP	6
```
Total nit-pick so feel free to ignore, but I'd prefer `*_SHIFT` rather than `*_MOVE_STEP`.

rriveramcrus commented on drivers/charger/charger_bq2562x.c:
```diff
@@ -0,0 +1,1035 @@
+	const struct bq2562x_config *const config = dev->config;
+	uint8_t ilim[2] = {0};
+
+	iindpm = CLAMP(iindpm, BQ2562X_IINDPM_I_MIN_uA,
+						BQ2562X_IINDPM_I_MAX_uA);
```
Indentation looks a little strange here, have you run `clang-format`?

rriveramcrus commented on include/zephyr/drivers/charger.h:
```diff
@@ -36,6 +36,8 @@ enum charger_property {
 	/** Value should be of type enum charger_online */
 	CHARGER_PROP_ONLINE = 0,
+	/** Represents the manufacturer of the charger */
+	CHARGER_PROP_MANUFACTURER,
```
https://sbs-forum.org/specs/sbdat110.pdf

So historically this property comes from the Linux's power supply class and its implementation of the SBS specification. The SBS specification defines `ManufacturerName()` as a field in a standardized register map on a fuel gauge IC integrated into a battery pack. 

These battery packs can have the same electrical characteristics and SKUs, but ultimately contain different fuel gauge ICs for logistical reasons and we use `ManufacturerName()` and other properties in the specification to identify the fuel gauge and handle quirks from a random SKU-compatible battery pack.

The SBS specification does not truly define a standard charger IC register map with the exception of the `ChargingCurrent()` and `ChargingVoltage()` fields for the purpose of allowing the Smart Battery of SMBus Host to configure the charger via SMBus message.

In Zephyr we have decoupled chargers and fuel gauges into separate APIs, and charging ICs are typically not hot-swappable the way fuel gauges are because charging ICs *almost* always live on system. For these reasons I am hesitant to introduce this property unless you can elaborate on the benefit/application of this property to the mainline kernel.

rriveramcrus commented on include/zephyr/drivers/charger.h:
```diff
@@ -267,6 +269,8 @@ union charger_propval {
 	/** CHARGER_PROP_ONLINE */
 	enum charger_online online;
+	/** CHARGER_PROP_MANUFACTURER */
+	const char *manufacturer;
```
Also if we do decide to continue with this feature, please have a look at `fuel_gauge.h` and let's implement it in a similar way.

rriveramcrus commented on include/zephyr/drivers/charger.h:
```diff
@@ -38,6 +38,8 @@ enum charger_property {
 	/** Represents the manufacturer of the charger */
 	CHARGER_PROP_MANUFACTURER,
+	/** Represents the model name of the charger */
+	CHARGER_PROP_MODEL_NAME,
```
My comment https://github.com/zephyrproject-rtos/zephyr/pull/87981/commits/a7eb9accd487d78db3f278f53d309627808ad974#r2035697859 applies for this property as well

rriveramcrus commented on include/zephyr/drivers/charger.h:
```diff
@@ -271,6 +273,8 @@ union charger_propval {
 	/** CHARGER_PROP_MANUFACTURER */
 	const char *manufacturer;
+	/** CHARGER_PROP_MODEL_NAME */
+	const char *model;
```
Similarly this comment https://github.com/zephyrproject-rtos/zephyr/pull/87981/commits/a7eb9accd487d78db3f278f53d309627808ad974#r2035700106 applies too.

rriveramcrus commented on include/zephyr/drivers/charger.h:
```diff
@@ -311,6 +315,10 @@ union charger_propval {
 	uint32_t charge_term_current_ua;
 	/** CHARGER_PROP_CONSTANT_CHARGE_VOLTAGE_UV */
 	uint32_t const_charge_voltage_uv;
+	/** CHARGER_PROP_BATTERY_VOLTAGE_NOW */
+	uint32_t bat_voltage_now_uv;
+	/** CHARGER_PROP_BATTERY_CURRENT_NOW */
+	int32_t bat_current_now_ua;
```
Nice work on this change, I like it. I'd prefer spelling out `battery_` instead of `bat_` for clarity.

rriveramcrus commented on include/zephyr/drivers/charger.h:
```diff
@@ -66,6 +66,10 @@ enum charger_property {
 	CHARGER_PROP_BATTERY_VOLTAGE_NOW,
 	/** Present battery current in uA */
 	CHARGER_PROP_BATTERY_CURRENT_NOW,
+	/** Present charger voltage in uV */
+	CHARGER_PROP_VOLTAGE_NOW,
+	/** Present charger current in uA */
+	CHARGER_PROP_CURRENT_NOW,
```
Might be worth calling the properties `CHARGER_PROP_INPUT_VOLTAGE_NOW` and `CHARGER_PROP_INPUT_CURRENT_NOW`

parthitce commented:
@rriveramcrus Thanks for your review and inputs. Addressed all the them.

parthitce committed:
dts: bindings: add BQ25620 and BQ25622 support

add BQ25620 and BQ25622 support

Signed-off-by: Parthiban Nallathambi <parthiban@linumiz.com>

parthitce committed:
charger: add support for TI's BQ25620

add support for TI's BQ25620. Although BQ25622 is same with additional
feature sets, which isn't currently tested with this changes.

Signed-off-by: Parthiban Nallathambi <parthiban@linumiz.com>

rriveramcrus commented on drivers/charger/charger_bq2562x.c:
```diff
@@ -0,0 +1,987 @@
+
+	charger_enable = chrg_ctrl_0 & BQ2562X_CHRG_EN;
+	if (charger_enable != 0) {
+		if (config->ce_gpio.port && !ce_pin) {
```
I don't like that the `ce_gpio` is mandatory. I'd imagine some users may be okay with enabling/disabling the charge cycle via the register map only.  Also `ce_pin` is inverted again?

parthitce replied:
Invert it wrong, but the check against not of NULL will help not mandate the CE presence.

rriveramcrus replied:
Stepping through this code.

First we initialize `ce_pin` to 1. Then if we have a `ce_gpio` we get its real value and assign it to `ce_pin`.

Next we read the `CHRG_EN` bit and assign it to `charger_enable`

*Finally, if `charger_enable` is set, then we logically AND whether or not we have a `ce_gpio` with `ce_pin`.*

Let's assume we don't have a `ce_gpio`. That would resolve to:
```c
if (NULL && 1) {
    return true;
}
```

Shouldn't it be
```c
if (ce_pin)
```

rriveramcrus commented on drivers/charger/charger_bq2562x.c:
```diff
@@ -0,0 +1,987 @@
+	return 0;
+}
+
+static int bq2562x_set_ichrg_curr(const struct device *dev, int chrg_curr)
```
`chrg_curr` should be `uint32_t`

parthitce replied:
Fixed, thanks

rriveramcrus commented on drivers/charger/charger_bq2562x.c:
```diff
@@ -0,0 +1,987 @@
+{
+	const struct bq2562x_config *const config = dev->config;
+	struct bq2562x_data *data = dev->data;
+	int chrg_curr_max = data->constant_charge_current_max_ua;
```
similarly `chrg_curr_max` should be `uint32_t`

parthitce replied:
Fixed, thanks

schuele-embedded commented on dts/bindings/charger/ti,bq2562x.yaml:
```diff
@@ -0,0 +1,105 @@
+    reg = <0x6b>;
+
+    ce-gpios = <&gpioa 3 GPIO_ACTIVE_LOW>;
+    watchdog-timeout-ms = <50000>;
```
Is actually not part of properties

parthitce replied:
removed, thanks

schuele-embedded commented on dts/bindings/charger/ti,bq2562x.yaml:
```diff
@@ -0,0 +1,105 @@
+    constant-charge-voltage-max-microvolt = <4200000>;
+    precharge-current-microamp = <100000>;
+    charge-term-current-microamp = <60000>;
+    input-current-limit-microamp = <1000000>;
```
Is listed in properties with vendor prefix "ti" as well as the following property. Please remove vendor prefix or adjust this sample.

tmon-nordic commented:
> > > Added the USB label here to indicate that there may be some discussion here related to how to implement the BC1.2 functionality of this chip.
> > 
> > @tmon-nordic are you the right person to comment on this? This chip does perform BC1.2 detection via D+/D-.
> 
> @parthitce, I suggest getting this in front of the TSC if reading the USB detection result is a must-have for your application. I have concerns that introducing this in the charger API as it duplicates functionality already in Zephyr.

Sorry for not responding earlier, I didn't look into the PR because I thought it was just some followups to #86791 and not a separate PR.

Currently the USB stack is not really coupled together with BC1.2 detection result. There is https://github.com/zephyrproject-rtos/zephyr/blob/5601af34b71e1108ee2be646a44327b51ff8ec07/include/zephyr/drivers/usb/usb_bc12.h#L101 that was introduced by @keith-zephyr which is pretty much independent from the USB device stack.

keith-zephyr commented:
>Thanks @tmon-nordic. Beyond just reporting the BC1.2 detection results it looks like the bc12_driver_api would allow us to support the BQ25620's reverse mode (aka OTG) by updating the bc12_role from BC12_DISCONNECTED or BC12_PORTABLE_DEVICE to BC12_CHARGING_PORT. @keith-zephyr any concerns?

If the BQ25620 can detect BC1.2, it should be straight forward to add an entry to the `bc12_driver_api` specific to the BQ25620.  There is a [unit test](https://github.com/zephyrproject-rtos/zephyr/blob/11c543e24cddeab1b393fbbec88c5517cb4d4456/tests/drivers/usb/bc12/src/charging_mode.c#L55) that demonstrates using `bc12_set_role()` to change the role to charnging mode.

parthitce commented on drivers/charger/charger_bq2562x.c:
```diff
@@ -0,0 +1,987 @@
+static int bq2562x_get_ichrg_curr(const struct device *dev, uint32_t *current_ua)
+{
+	const struct bq2562x_config *const config = dev->config;
+	uint8_t ichg[2] = {0};
```
Sure, removed it.

parthitce commented on drivers/charger/charger_bq2562x.c:
```diff
@@ -0,0 +1,987 @@
+	int ret;
+
+	if (config->ce_gpio.port != NULL) {
+		ce_pin = !gpio_pin_get_dt(&config->ce_gpio);
```
Thanks, invert is wrong. Fixed.

parthitce commented:
@rriveramcrus As the changes required additional MFD driver and wrappers to it, I have dropped the USB related changes now. Yes, we require USB detection functionality and also additional changes for this IC. We/I will address both of them in a followup PR. Please revisit. Thanks.

parthitce commented on drivers/charger/charger_bq2562x.c:
```diff
@@ -0,0 +1,987 @@
+		case BQ2562X_TRICKLE_CHRG:
+			*type = CHARGER_CHARGE_TYPE_TRICKLE;
+			ret = bq2562x_get_ibat_adc(dev, &ibat);
+			if (ret) {
+				break;
+			}
+
+			ret = i2c_reg_read_byte_dt(&config->i2c, BQ2562X_CHRG_CTRL_0, &chrg_ctl);
+			if (ret) {
+				break;
+			}
```
Sure, thanks for testing it.

dsseng replied:
I did not test, but now following the code lets me be sure errors are passed to the caller. No hardware to test on my side

rriveramcrus commented on drivers/charger/charger_bq2562x.c:
```diff
@@ -0,0 +1,947 @@
+
+	data->dev = dev;
+	ret = i2c_reg_read_byte_dt(&config->i2c, BQ2562X_PART_INFO, &val);
+	if (ret) {
```
https://gitlab.com/MISRA/MISRA-C/MISRA-C-2012/Example-Suite/-/blob/master/R_14_04.c#L44

https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html

This does not comply with the project coding guidelines.

rriveramcrus commented on drivers/charger/charger_bq2562x.c:
```diff
@@ -0,0 +1,947 @@
+	struct bq2562x_data *data = dev->data;
+	int ret;
+
+	ret = i2c_reg_update_byte_dt(&config->i2c, BQ2562X_CHRG_CTRL_2, BQ2562X_CTRL2_REG_RST,
```
Not a problem with your code, but something I remembered about this chip is that `REG_RESET` will flip `EN_CHG` to the Enable state. (_Table 8-20. REG0x16_Charger_Control_1 Register Field Descriptions_)

In the event a consumer of this driver is tying the enable pin low, the register reset action would result in a charging cycle commencing. This in itself is not a problem necessarily since the defaults are fairly safe, but the rest of this function configures some of the controls for the charging state machine.

I vaguely recall the charge voltage setting latches at the start of a charge cycle and subsequent writes to the control register have no effect until the charge cycle is toggled.

Might be a good idea to clear `EN_CHG` coming out of reset and then proceed with configuring the charger. Alternatively, check with TI or try it out on your hardware to see if my vague memory is still correct.

rriveramcrus commented on drivers/charger/charger_bq2562x.c:
```diff
@@ -0,0 +1,947 @@
+	case BQ2562X_TEMP_PREWARM:
+		*health = CHARGER_HEALTH_WARM;
+		break;
+	case BQ2562X_TEMP_PIN_BIAS_REF_FAULT:
```
Isn't this only a BQ25622 feature? This driver doesn't support BQ25622 according to line 874

parthitce commented:
Removed stale. I will resume this by end of November.
